File: Provenance/Controller/PVAtari5200ControllerViewController.m
//
//  PVAtari5200ControllerViewController.m
//  Provenance
//
//  Created by Joe Mattiello on 11/28/2016.
//  Copyright (c) 2013 James Addyman. All rights reserved.
//

#import "PVAtari5200ControllerViewController.h"
#import <PVAtari800/PVAtari800.h>

@interface PVAtari5200ControllerViewController ()

@end

@implementation PVAtari5200ControllerViewController

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    
    for (JSButton *button in [self.buttonGroup subviews])
    {
        if (![button isMemberOfClass:[JSButton class]])
        {
            continue; // skip over the PVButtonGroupOverlayView
        }
        
        if ([[[button titleLabel] text] isEqualToString:@"Fire 1"]) {
            [button setTag:PV5200ButtonFire1];
        } else  if ([[[button titleLabel] text] isEqualToString:@"Fire 2"]) {
            [button setTag:PV5200ButtonFire2];
        }
    }
    
    self.leftShoulderButton.tag = PV5200ButtonReset;
    
    self.startButton.tag  = PV5200ButtonStart;
    self.selectButton.tag = PV5200ButtonPause;
}

- (void)dPad:(JSDPad *)dPad didPressDirection:(JSDPadDirection)direction
{
    ATR800GameCore *a800SystemCore = (ATR800GameCore *)self.emulatorCore;
    
    [a800SystemCore didRelease5200Button:PV5200ButtonUp forPlayer:0];
    [a800SystemCore didRelease5200Button:PV5200ButtonDown forPlayer:0];
    [a800SystemCore didRelease5200Button:PV5200ButtonLeft forPlayer:0];
    [a800SystemCore didRelease5200Button:PV5200ButtonRight forPlayer:0];
    
    switch (direction)
    {
        case JSDPadDirectionUpLeft:
            [a800SystemCore didPush5200Button:PV5200ButtonUp forPlayer:0];
            [a800SystemCore didPush5200Button:PV5200ButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionUp:
            [a800SystemCore didPush5200Button:PV5200ButtonUp forPlayer:0];
            break;
        case JSDPadDirectionUpRight:
            [a800SystemCore didPush5200Button:PV5200ButtonUp forPlayer:0];
            [a800SystemCore didPush5200Button:PV5200ButtonRight forPlayer:0];
            break;
        case JSDPadDirectionLeft:
            [a800SystemCore didPush5200Button:PV5200ButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionRight:
            [a800SystemCore didPush5200Button:PV5200ButtonRight forPlayer:0];
            break;
        case JSDPadDirectionDownLeft:
            [a800SystemCore didPush5200Button:PV5200ButtonDown forPlayer:0];
            [a800SystemCore didPush5200Button:PV5200ButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionDown:
            [a800SystemCore didPush5200Button:PV5200ButtonDown forPlayer:0];
            break;
        case JSDPadDirectionDownRight:
            [a800SystemCore didPush5200Button:PV5200ButtonDown forPlayer:0];
            [a800SystemCore didPush5200Button:PV5200ButtonRight forPlayer:0];
            break;
        default:
            break;
    }
    
    [self vibrate];
}

- (void)dPadDidReleaseDirection:(JSDPad *)dPad
{
    ATR800GameCore *a800SystemCore = (ATR800GameCore *)self.emulatorCore;
    
    [a800SystemCore didRelease5200Button:PV5200ButtonUp forPlayer:0];
    [a800SystemCore didRelease5200Button:PV5200ButtonDown forPlayer:0];
    [a800SystemCore didRelease5200Button:PV5200ButtonLeft forPlayer:0];
    [a800SystemCore didRelease5200Button:PV5200ButtonRight forPlayer:0];
}

- (void)buttonPressed:(JSButton *)button
{
    ATR800GameCore *a800SystemCore = (ATR800GameCore *)self.emulatorCore;
    PV5200Button tag = button.tag;
    [a800SystemCore didPush5200Button:tag forPlayer:0];
    [self vibrate];
}

- (void)buttonReleased:(JSButton *)button
{
    ATR800GameCore *a800SystemCore = (ATR800GameCore *)self.emulatorCore;
    PV5200Button tag = button.tag;
    [a800SystemCore didRelease5200Button:tag forPlayer:0];
}

- (void)pressStartForPlayer:(NSUInteger)player
{
    ATR800GameCore *a800SystemCore = (ATR800GameCore *)self.emulatorCore;
    [a800SystemCore didPush5200Button:PV5200ButtonReset forPlayer:player];
}

- (void)releaseStartForPlayer:(NSUInteger)player
{
    ATR800GameCore *a800SystemCore = (ATR800GameCore *)self.emulatorCore;
    [a800SystemCore didRelease5200Button:PV5200ButtonReset forPlayer:player];
}

- (void)pressSelectForPlayer:(NSUInteger)player
{
    ATR800GameCore *a800SystemCore = (ATR800GameCore *)self.emulatorCore;
    [a800SystemCore didPush5200Button:PV5200ButtonPause forPlayer:player];
}

- (void)releaseSelectForPlayer:(NSUInteger)player
{
    ATR800GameCore *a800SystemCore = (ATR800GameCore *)self.emulatorCore;
    [a800SystemCore didRelease5200Button:PV5200ButtonPause forPlayer:player];
}

@end

File: Provenance/Controller/PVAtari5200ControllerViewController.swift
//
//  PVAtari5200ControllerViewController.swift
//  Provenance
//
//  Created by Joe Mattiello on 17/03/2018.
//  Copyright (c) 2018 Joe Mattiello. All rights reserved.
//

import PVSupport

fileprivate extension JSButton {
    var buttonTag: PV5200Button {
        get {
            return PV5200Button(rawValue: tag)!
        }
        set {
            tag = newValue.rawValue
        }
    }
}

class PVAtari5200ControllerViewController: PVControllerViewController<PV5200SystemResponderClient> {

    override func layoutViews() {
        buttonGroup?.subviews.forEach {
            guard let button = $0 as? JSButton, let title = button.titleLabel?.text else {
                return
            }
            if title == "Fire 1" || title == "1" {
                button.buttonTag = .fire1
            } else if title == "Fire 2" || title == "2" {
                button.buttonTag = .fire2
            }
        }

        leftShoulderButton?.buttonTag = .reset
        startButton?.buttonTag = .start
        selectButton?.buttonTag = .pause
    }

    override func dPad(_ dPad: JSDPad, didPress direction: JSDPadDirection) {
        emulatorCore.didRelease(.up, forPlayer: 0)
        emulatorCore.didRelease(.down, forPlayer: 0)
        emulatorCore.didRelease(.left, forPlayer: 0)
        emulatorCore.didRelease(.right, forPlayer: 0)
        switch direction {
            case .upLeft:
                emulatorCore.didPush(.up, forPlayer: 0)
                emulatorCore.didPush(.left, forPlayer: 0)
            case .up:
                emulatorCore.didPush(.up, forPlayer: 0)
            case .upRight:
                emulatorCore.didPush(.up, forPlayer: 0)
                emulatorCore.didPush(.right, forPlayer: 0)
            case .left:
                emulatorCore.didPush(.left, forPlayer: 0)
            case .right:
                emulatorCore.didPush(.right, forPlayer: 0)
            case .downLeft:
                emulatorCore.didPush(.down, forPlayer: 0)
                emulatorCore.didPush(.left, forPlayer: 0)
            case .down:
                emulatorCore.didPush(.down, forPlayer: 0)
            case .downRight:
                emulatorCore.didPush(.down, forPlayer: 0)
                emulatorCore.didPush(.right, forPlayer: 0)
            default:
                break
        }
        vibrate()
    }

    override func dPadDidReleaseDirection(_ dPad: JSDPad) {
        emulatorCore.didRelease(.up, forPlayer: 0)
        emulatorCore.didRelease(.down, forPlayer: 0)
        emulatorCore.didRelease(.left, forPlayer: 0)
        emulatorCore.didRelease(.right, forPlayer: 0)
    }

    override func buttonPressed(_ button: JSButton) {
        emulatorCore.didPush(button.buttonTag, forPlayer: 0)
        vibrate()
    }

    override func buttonReleased(_ button: JSButton) {
        emulatorCore.didRelease(button.buttonTag, forPlayer: 0)
    }

    override func pressStart(forPlayer player: Int) {
        emulatorCore.didPush(.reset, forPlayer: player)
    }

    override func releaseStart(forPlayer player: Int) {
        emulatorCore.didRelease(.reset, forPlayer: player)
    }

    override func pressSelect(forPlayer player: Int) {
        emulatorCore.didPush(.pause, forPlayer: player)
    }

    override func releaseSelect(forPlayer player: Int) {
        emulatorCore.didRelease(.pause, forPlayer: player)
    }
}
